test/utils.dart

```diff
@@ -25,17 +25,89 @@ const formattedOutput = 'void main() => print("hello");';
 final _indentPattern = RegExp(r'\(indent (\d+)\)');
 final _fixPattern = RegExp(r'\(fix ([a-x-]+)\)');
 
-/// Runs the command line formatter, passing it [args].
-Future<TestProcess> runFormatter([List<String> args]) {
+/// If tool/command_shell.dart has been compiled to a snapshot, this is the path
+/// to it.
+String _commandExecutablePath;
+
+/// If bin/format.dart has been compiled to a snapshot, this is the path to it.
+String _formatterExecutablePath;
+
+/// Compiles format.dart to a native executable for tests to use.
+void compileFormatterExecutable() {
+  setUpAll(() async {
+    _formatterExecutablePath = await _compileExecutable('bin/format.dart');
+  });
+
+  tearDownAll(() async {
+    await _deleteExecutable(_formatterExecutablePath);
+    _formatterExecutablePath = null;
+  });
+}
+
+/// Compiles command_shell.dart to a native executable for tests to use.
+void compileCommandExecutable() {
+  setUpAll(() async {
+    _commandExecutablePath =
+        await _compileExecutable('tool/command_shell.dart');
+  });
+
+  tearDownAll(() async {
+    await _deleteExecutable(_commandExecutablePath);
+    _commandExecutablePath = null;
+  });
+}
+
+/// Compile the Dart [script] to a native executable.
+///
+/// We do this instead of spawning the script from source each time because it's
+/// much faster when the same script needs to be run several times.
+Future<String> _compileExecutable(String script) async {
+  var scriptName = p.basename(script);
+  var tempDir =
+      await Directory.systemTemp.createTemp(p.withoutExtension(scriptName));
+  var executable = p.join(tempDir.path, '$scriptName.aot');
+
+  var dart2Native =
```

Comment:
Have you compared this against `dart --snapshot` for overall runtime?

Reply:
I didn't. I thought about it, but the docs for the old kernel snapshots sound like it wouldn't offer much speed improvement, and the app-JIT snapshots require some kind of training data, which seemed like more trouble than it was worth.

But now you have piqued my curiosity. I tried all three options:

AOT:

```
00:42 +1801: All tests passed!
real  0m43.214s
user  1m17.798s
sys 0m7.271s
```

Kernel snapshot:

```
00:39 +1801: All tests passed!
real  0m40.140s
user  1m15.609s
sys 0m9.052s
```

App JIT snapshot (trained using just `--help`):

```
00:36 +1801: All tests passed!
real  0m36.581s
user  1m12.847s
sys 0m8.341s
```

So changed it to use the latter instead.